src/agent/src/device.rs

```diff
@@ -436,217 +426,242 @@ fn scan_scsi_bus(scsi_addr: &str) -> Result<()> {
+        let guest_minor = stat::minor(devid);
+
+        Ok(SpecUpdate {
+            guest_major: guest_major as i64,
+            guest_minor: guest_minor as i64,
+            final_path: None,
+            pci: Vec::new(),
+        })
+    }
+    fn with_final_and_pci<T: AsRef<Path>>(
+        vm_path: T,
+        final_path: String,
+        pci: Vec<(pci::Address, pci::Address)>,
+    ) -> Result<Self> {
+        Ok(SpecUpdate {
+            final_path: Some(final_path),
+            pci,
+            ..Self::from_vm_path(vm_path)?
+        })
     }
+}
 
+// update_spec_devices updates the device list in the OCI spec to make
+// it include details appropriate for the VM, instead of the host.  It
+// is given a map of (container_path => update) where:
+//     container_path: the path to the device in the original OCI spec
+//     update: information on changes to make to the device
+#[instrument]
+fn update_spec_devices(spec: &mut Spec, mut updates: HashMap<&str, SpecUpdate>) -> Result<()> {
     let linux = spec
         .linux
         .as_mut()
-        .ok_or_else(|| anyhow!("Spec didn't container linux field"))?;
+        .ok_or_else(|| anyhow!("Spec didn't contain linux field"))?;
+    let mut resupdates = HashMap::<(&str, i64, i64), SpecUpdate>::with_capacity(updates.len());
+    let mut pcimap = HashMap::<pci::Address, pci::Address>::new();
+
+    for specdev in &mut linux.devices {
+        if let Some(mut update) = updates.remove(specdev.path.as_str()) {
+            let host_major = specdev.major;
+            let host_minor = specdev.minor;
 
-    if !Path::new(vm_path).exists() {
-        return Err(anyhow!("vm_path:{} doesn't exist", vm_path));
+            info!(
+                sl!(),
+                "update_spec_devices() updating device";
+                "container_path" => &specdev.path,
+                "type" => &specdev.r#type,
+                "host_major" => host_major,
+                "host_minor" => host_minor,
+                "guest_major" => update.guest_major,
+                "guest_minor" => update.guest_minor,
+                "final_path" => update.final_path.as_ref(),
+            );
+
+            specdev.major = update.guest_major;
+            specdev.minor = update.guest_minor;
+            if let Some(final_path) = update.final_path.take() {
+                specdev.path = final_path;
+            }
+
+            for (host, guest) in update.pci.drain(..) {
+                if let Some(other_guest) = pcimap.insert(host, guest) {
+                    return Err(anyhow!(
+                        "Conflicting guest address for host device {} ({} versus {})",
+                        host,
+                        guest,
+                        other_guest
+                    ));
+                }
+            }
+
+            if resupdates
+                .insert((&specdev.r#type, host_major, host_minor), update)
+                .is_some()
+            {
+                return Err(anyhow!(
```

Comment:
Worth adding `specdev` to the returned error?

Reply:
Eh, sure, updated.

Reply:
Oof, wait, I take that back.  At least when combined with splitting out the single device update into a helper function, attempting to print `specdev` here results in a horrible lifetime tangle (the fact that the resource map key points into the oci spec for the `type` field keeps the mutable reference to specdev alive overlapping with the immutable reference we'd need to print it here).

I tried a few ways to work around it, but didn't succeed.  I'm sure it's possible, but not worth the effort for just making an unlikely error message marginally nicer.